src/scaffoldmaker/utils/cylindermesh.py

```diff
@@ -0,0 +1,415 @@
+"""
+Utility functions for generating a generalised 3-D solid cylinder (extruded ellipse/circle). It can be used to genrate a
+solid truncated cone. It also can be used for transition from a 2D base to another base (e.g., ellipse to a circle).
+"""
+
+from enum import Enum
+from scaffoldmaker.utils import vector, geometry, mirror
+import math
+from opencmiss.zinc.field import Field
+from opencmiss.zinc.node import Node
+from opencmiss.utils.zinc.finiteelement import getMaximumNodeIdentifier, getMaximumElementIdentifier
+from scaffoldmaker.utils.shieldmesh import ShieldMesh, ShieldMode, SheildType
+from scaffoldmaker.utils.interpolation import sampleCubicHermiteCurves, interpolateSampleCubicHermite,\
+    smoothCubicHermiteDerivativesLine
+
+class CylinderMode(Enum):
+    CYLINDER_MODE_FULL = 1  # full cylinder is createted
+    CYLINDER_MODE_LOWER_HALF = 2  # lower half cylinder
+
+class CylinderType(Enum):
+    CYLIDNER_REGULAR = 1        # all the bases along the axis of the cylinder are the same.
+    CYLIDNER_TRUNCATED_CONE = 2 # different ellipses along the cylinder axis
+
+class ConeBaseProgression(Enum):
+    GEOMETIRC_PROGRESSION  = 1 # geometirc sequence decrease for major radius of bases
+    ARITHMETIC_PROGRESSION = 2 # arithmetic sequence decrease for major radius of bases
+
+class CylinderMesh:
+    '''
+    Cylinder mesh generator. Extrudes an ellipse/circle.
+    '''
+
+    def __init__(self,fieldModule, coordinates, baseCentre, alongAxis, majorAxis, minorRadius,
+                             elementsCountAcross, elementsCountUp, elementsCountAlong,
+                             cylinderMode = CylinderMode.CYLINDER_MODE_FULL,
+                             cylinderType = CylinderType.CYLIDNER_REGULAR,
+                             rate = 0.0, progressionMode=ConeBaseProgression.GEOMETIRC_PROGRESSION, useCrossDerivatives = False):
```

Comment:
What do you mean "provide a central path with major/minor radius fields"?

Reply:
I defined a class to store these parameters in it. I called it TAPERED and I store major and minor common ratio and progression type (r_n+1=r_n+ratio for arithmetic and r_n+1=r_n*ratio for geometric progression). In case of geometric sequence "ratio" makes sense but for arithmetic it should be common difference. I call it ratio and there is a comment to explain this. Do you have any better terms for TAPERED class or ratio?